Nightly integration failure: GitLab.com

Nightly integration tests failed for `gitlab-saas`.

| Field | Value |
|---|---|
| Adapter | `gitlab-saas` (tracker) |
| Tested version | remote service (no pinned client) |
| Install source | GitLab.com REST API v4 (remote service, no pinned version) |
| Run | https://github.com/sortie-ai/sortie/actions/runs/32094467935 |
| Commit | `0405e1290024b182ba1ba0a9c9a03de2ce072711` |
| Date | 2026-08-18 03:12 UTC |

### Failing tests

- `TestIntegration_FetchIssueStatesByIDs`
- `TestIntegration_SCMMergeFlow`
- `TestIntegration_SCMMergeFlow/MergePR_StalePrecondition`

### Log excerpt

```
TestIntegration_SCMMergeFlow/MergePR_StalePrecondition
    integration_merge_test.go:397: POST /projects/sergheii%2Fsortie-tracker-lab/repository/commits: Post "https://gitlab.com/api/v4/projects/sergheii%2Fsortie-tracker-lab/repository/commits": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
--- FAIL: TestIntegration_SCMMergeFlow/MergePR_StalePrecondition (30.00s)
    integration_merge_test.go:434: MergePR_StalePrecondition subtest failed; skipping dependent subtests
--- FAIL: TestIntegration_SCMMergeFlow (35.44s)
=== RUN   TestIntegration_FetchPendingReviews
    integration_scm_test.go:111: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchPendingReviews (0.00s)
=== RUN   TestIntegration_FetchBotReviewComments
    integration_scm_test.go:143: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchBotReviewComments (0.00s)
=== RUN   TestIntegration_GetReviewDecision
    integration_scm_test.go:209: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_GetReviewDecision (0.00s)
=== RUN   TestIntegration_GetMergeability
    integration_scm_test.go:243: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_GetMergeability (0.00s)
=== RUN   TestIntegration_GetCIStatus
    integration_scm_test.go:333: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_GetCIStatus (0.00s)
=== RUN   TestIntegration_FetchCIStatus
    integration_scm_test.go:365: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchCIStatus (0.00s)
=== RUN   TestIntegration_FetchCIStatus_ManyStatuses
    integration_scm_test.go:417: optional environment variable SORTIE_GITLAB_MANY_STATUS_SHA is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchCIStatus_ManyStatuses (0.00s)
=== RUN   TestIntegration_ListLabelEvents
    integration_scm_test.go:462: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_ListLabelEvents (0.00s)
=== RUN   TestIntegration_RemoveLabel
    integration_scm_test.go:506: optional environment variable SORTIE_GITLAB_LABEL_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_RemoveLabel (0.00s)
=== RUN   TestIntegration_VerifyAutoMergeScopes
    integration_scm_test.go:537: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_VerifyAutoMergeScopes (0.00s)
=== RUN   TestIntegration_SCMErrorKinds
    integration_scm_test.go:560: optional environment variable SORTIE_GITLAB_MR_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_SCMErrorKinds (0.00s)
=== RUN   TestIntegration_FetchCandidateIssues
--- PASS: TestIntegration_FetchCandidateIssues (0.99s)
=== RUN   TestIntegration_FetchCandidateIssues_FixtureFloor
--- PASS: TestIntegration_FetchCandidateIssues_FixtureFloor (1.21s)
=== RUN   TestIntegration_FetchIssueByID
--- PASS: TestIntegration_FetchIssueByID (1.85s)
=== RUN   TestIntegration_FetchIssueByID_NotFound
--- PASS: TestIntegration_FetchIssueByID_NotFound (1.30s)
=== RUN   TestIntegration_FetchIssueByID_WorkItem
    integration_test.go:253: optional environment variable SORTIE_GITLAB_TASK_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchIssueByID_WorkItem (0.00s)
=== RUN   TestIntegration_FetchIssuesByStates_Empty
--- PASS: TestIntegration_FetchIssuesByStates_Empty (0.65s)
=== RUN   TestIntegration_FetchIssuesByStates_Active
--- PASS: TestIntegration_FetchIssuesByStates_Active (1.19s)
=== RUN   TestIntegration_FetchIssuesByStates_Terminal
--- PASS: TestIntegration_FetchIssuesByStates_Terminal (1.19s)
=== RUN   TestIntegration_FetchIssueStatesByIDs
    integration_test.go:395: FetchIssueStatesByIDs: context deadline exceeded
--- FAIL: TestIntegration_FetchIssueStatesByIDs (30.81s)
=== RUN   TestIntegration_FetchIssueStatesByIdentifiers
--- PASS: TestIntegration_FetchIssueStatesByIdentifiers (1.17s)
=== RUN   TestIntegration_FetchIssueComments
--- PASS: TestIntegration_FetchIssueComments (1.56s)
=== RUN   TestIntegration_FetchIssueComments_SystemNotesExcluded
    integration_test.go:496: optional environment variable SORTIE_GITLAB_SYSTEM_NOTES_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchIssueComments_SystemNotesExcluded (0.00s)
=== RUN   TestIntegration_FetchIssueComments_Paginated
    integration_test.go:518: optional environment variable SORTIE_GITLAB_MANY_NOTES_IID is not set; skipping the fixture it gates
--- SKIP: TestIntegration_FetchIssueComments_Paginated (0.00s)
=== RUN   TestIntegration_TransitionIssue_SameState
--- PASS: TestIntegration_TransitionIssue_SameState (1.69s)
=== RUN   TestIntegration_TransitionIssue_RoundTrip
--- PASS: TestIntegration_TransitionIssue_RoundTrip (3.82s)
=== RUN   TestIntegration_TransitionIssue_TerminalAndBack
--- PASS: TestIntegration_TransitionIssue_TerminalAndBack (5.63s)
=== RUN   TestIntegration_TransitionIssue_InvalidTarget
--- PASS: TestIntegration_TransitionIssue_InvalidTarget (0.83s)
=== RUN   TestIntegration_CommentIssue
--- PASS: TestIntegration_CommentIssue (2.23s)
=== RUN   TestIntegration_AddLabel
--- PASS: TestIntegration_AddLabel (2.09s)
=== RUN   TestIntegration_NewGitLabAdapter_ForeignProject
    integration_test.go:789: optional environment variable SORTIE_GITLAB_FOREIGN_PROJECT is not set; skipping the fixture it gates
--- SKIP: TestIntegration_NewGitLabAdapter_ForeignProject (0.00s)
=== RUN   TestIntegration_FetchCandidateIssues_QueryFilter
--- PASS: TestIntegration_FetchCandidateIssues_QueryFilter (1.68s)
FAIL
FAIL	github.com/sortie-ai/sortie/internal/scm/gitlab	95.350s
```

_Filed by the nightly integration monitor. It comments on recurrence and closes automatically when this adapter passes again._

--- a/docs/gitlab-adapter-notes.md
+++ b/docs/gitlab-adapter-notes.md
@@ -1383,6 +1383,9 @@ load-bearing rather than incidental:
   `rel="next"` link, with exactly 100 entries on page one, because the adapter sends
   `per_page=100` **[live-CE]**) to assert note pagination, and one non-`issue` work item to
   assert the `issue_type=issue` exclusion.
+- A merge-flow fixture that asserts on a branch after merging its merge request opens that
+  merge request with source-branch removal disabled: the governing project setting defaults to
+  removal, and the removal is asynchronous, so a fixture that leaves it enabled races the reap.
 
 ---
 
@@ -2188,11 +2191,29 @@ which is what `adaptertest.AssertBranchAbsentDisposition` pins. `RemoveLabel` di
 already-absent case the other way, returning nil
 (`adaptertest.AssertLabelAbsentDisposition`), so the two must not be written from one template.
 Branch names containing a slash MUST be percent-encoded: `feat%2Fnested-name` deleted
-successfully **[live-CE]**. Note that `should_remove_source_branch: true` on the merge call is
-asynchronous; the branch was still readable immediately after a 200 merge response
-**[live-CE]**, so `DeleteBranch` must tolerate both the branch being present and it having
-already been reaped. Neither route may produce `ErrSCMConflict`, whatever status it returns:
-only the merge write path promotes.
+successfully **[live-CE]**. The reap is asynchronous, and it is governed by the merge request's
+own `force_remove_source_branch` rather than by any field the merge call itself sends: the branch
+was still readable immediately after a 200 merge response **[live-CE]**, so `DeleteBranch` must
+tolerate both the branch being present and it having already been reaped. `force_remove_source_branch`
+is resolved at creation from the project's `remove_source_branch_after_merge` when the create
+request omits `remove_source_branch`; the adapter's own merge call sends no field that influences
+it. `remove_source_branch_after_merge` is `true` by default: a project created with no
+merge-behavior field set on the pinned Community Edition 19.2.1 image carries it **[live-CE]**,
+and it is also `true` on the GitLab.com lab project **[live-SaaS]** and on the project
+`scripts/gitlab-integration-provision.sh` creates **[live-CE]**. The reap window is variable and
+can fall below a client's own post-merge round-trip cost, which is what makes a delete-after-merge
+assertion a race rather than a sequence: the branch survived 0.739 s after a squash merge on
+GitLab.com **[live-SaaS]**, 0.513 s and 2.011 s in two trials on the self-managed instance
+**[live-CE]**, and 1.123 s on the provisioned container **[live-CE]**; these windows are scoped to
+the deployments where they were observed and are not generalized to any deployment not measured.
+Sending `remove_source_branch: false` on `POST /projects/:id/merge_requests` overrides the project
+setting: the create response echoes `force_remove_source_branch: false`, and the branch then
+survived for the whole observation window on both deployments, 20 s (58 polls) on the self-managed
+instance and 25 s (40 polls) on GitLab.com **[live-CE]** **[live-SaaS]**. A branch the forge
+reaped and a branch the caller deleted are indistinguishable at this route: both answer a
+subsequent `DELETE` with `404 {"message":"404 Branch Not Found"}`, byte-identical, on both
+deployments **[live-CE]** **[live-SaaS]**. Neither route may produce `ErrSCMConflict`, whatever
+status it returns: only the merge write path promotes.
 
 ### 7. Bot classification
 

--- a/docs/gitlab-adapter-notes.md
+++ b/docs/gitlab-adapter-notes.md
@@ -2209,7 +2209,7 @@ the deployments where they were observed and are not generalized to any deployme
 Sending `remove_source_branch: false` on `POST /projects/:id/merge_requests` overrides the project
 setting: the create response echoes `force_remove_source_branch: false`, and the branch then
 survived for the whole observation window on both deployments, 20 s (58 polls) on the self-managed
-instance and 25 s (40 polls) on GitLab.com **[live-CE]** **[live-SaaS]**. A branch the forge
+instance **[live-CE]** and 25 s (40 polls) on GitLab.com **[live-SaaS]**. A branch the forge
 reaped and a branch the caller deleted are indistinguishable at this route: both answer a
 subsequent `DELETE` with `404 {"message":"404 Branch Not Found"}`, byte-identical, on both
 deployments **[live-CE]** **[live-SaaS]**. Neither route may produce `ErrSCMConflict`, whatever

--- a/internal/scm/gitlab/integration_merge_test.go
+++ b/internal/scm/gitlab/integration_merge_test.go
@@ -161,19 +161,33 @@ func (f *gitlabMergeFixture) commitFile(t *testing.T, project, branch, path, con
 }
 
 // openMR opens a merge request from source to target and returns its
-// iid.
+// iid. It opens the merge request with source-branch removal disabled:
+// the project setting governing removal defaults to removing the
+// source branch, the removal is asynchronous, and a test that asserts
+// on the branch after the merge would otherwise be racing that reap.
+// Dropping this would silently reintroduce that race.
+//
+// It also decodes and checks the create response's own copy of the
+// setting, so a forge that ignores the request fails this setup step
+// rather than a later subtest that would misread the symptom as a
+// delete defect.
 func (f *gitlabMergeFixture) openMR(t *testing.T, project, source, target, title string) int {
 	t.Helper()
 	resp := f.do(t, http.MethodPost, "/projects/"+project+"/merge_requests", map[string]any{
-		"source_branch": source,
-		"target_branch": target,
-		"title":         title,
+		"source_branch":        source,
+		"target_branch":        target,
+		"title":                title,
+		"remove_source_branch": false,
 	})
 	var mr struct {
-		IID int `json:"iid"`
+		IID                     int  `json:"iid"`
+		ForceRemoveSourceBranch bool `json:"force_remove_source_branch"`
 	}
 	if err := json.Unmarshal(resp, &mr); err != nil {
 		t.Fatalf("unmarshal create merge request response: %v", err)
+	}
+	if mr.ForceRemoveSourceBranch {
+		t.Fatalf("merge request !%d force_remove_source_branch = %v, want %v", mr.IID, true, false)
 	}
 	return mr.IID
 }
@@ -484,6 +498,10 @@ func TestIntegration_SCMMergeFlow(t *testing.T) {
 		t.Skip("MergePR_AlreadyMerged subtest failed; skipping DeleteBranch subtests")
 	}
 
+	// head is still present here because openMR opened the merge request
+	// with source-branch removal disabled, so the forge did not reap it
+	// on merge. This subtest exercises the adapter's own delete, not the
+	// forge's asynchronous one.
 	t.Run("DeleteBranch", func(t *testing.T) {
 		if err := adapter.DeleteBranch(ctx, owner, repo, head); err != nil {
 			t.Fatalf("DeleteBranch(%q): %v", head, err)

